### ISSUE TITLE
Contour ingress isn't working

### ISSUE THREAD

**Issue by brandonh-msft** [kubernetes-sigs/kind#3181]
I went thru the Ingress documentation to set up Contour. I got to the bottom and ran the two `curl` commands and am not met with any echo "foo-app" or "bar-app" as it says should happen. Instead, I get:

```
> curl localhost:8100/bar/hostname
curl: (52) Empty reply from server
```

Note I have to use `:8100` because my kind cluster is being setup w/ a different template than the default shown in the walkthrough

<details>
<summary>my cluster template</summary>

```yaml
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
nodes:
- role: control-plane
  kubeadmConfigPatches:
  - |
    kind: InitConfiguration
    nodeRegistration:
      kubeletExtraArgs:
        node-labels: "ingress-ready=true"
  extraPortMappings:
  - containerPort: 80
    hostPort: 8100
    protocol: TCP
  - containerPort: 443
    hostPort: 8443
    protocol: TCP
```

</details>

And so I modified the "Using ingress" template to replace all the 8080 with 80, then attempt to hit 8100.
If i try to just hit 80 I get
```
> curl localhost/bar/hostname
curl: (7) Failed to connect to localhost port 80 after 2242 ms: Couldn't connect to server
```

as I (believe) is correct given my cluster setup.

I am wondering if the images being used no longer echo back and, instead, just ACK as a way to signify e2e is working? If so, it seems this doc needs an update.

**Comment by BenTheElder** [comment]
/remove-kind documentation
/kind support

> I am wondering if the images being used no longer echo back and, instead, just ACK as a way to signify e2e is working? If so, it seems this doc needs an update.

The image being used for the sample app definitely replies back.
I just confirmed using the ingress guide precisely and ingress-nginx on kind v0.18

Also, we maintain the image used for the ingress sample as well, it's used to e2e test kubernetes.
```
$ curl localhost/foo/hostname
foo-app
```

What exactly did you customize?

**Comment by BenTheElder** [comment]
You may be getting an empty reply from the ingress controller

**Comment by BenTheElder** [comment]
> my cluster template
> ```yaml kind: Cluster apiVersion: kind.x-k8s.io/v1alpha4 nodes: - role: control-plane kubeadmConfigPatches: - | kind: InitConfiguration nodeRegistration: kubeletExtraArgs: node-labels: "ingress-ready=true" extraPortMappings: - containerPort: 80 hostPort: 8100 protocol: TCP - containerPort: 443 hostPort: 8443 protocol: TCP ```

If this is the only aspect that was customized, I would say there's a problem with the contour controller then.

We should consider removing these, ingress-nginx is the only one really getting active support from the ingress maintainers. We recently removed ambassador for the same reason.

Let me try again here with the contour guide unmodified ...

**Comment by BenTheElder** [comment]
I get:
```
$ curl localhost/foo/hostname
curl: (56) Recv failure: Connection reset by peer
```

So I'd say the countour deployment is broken:
```
$ kubectl get po -A
NAMESPACE            NAME                                         READY   STATUS      RESTARTS   AGE
default              bar-app                                      1/1     Running     0          3m51s
default              foo-app                                      1/1     Running     0          3m51s
kube-system          coredns-787d4945fb-8lmfn                     1/1     Running     0          3m59s
kube-system          coredns-787d4945fb-xzqk4                     1/1     Running     0          3m59s
kube-system          etcd-kind-control-plane                      1/1     Running     0          4m13s
kube-system          kindnet-w4j6n                                1/1     Running     0          3m59s
kube-system          kube-apiserver-kind-control-plane            1/1     Running     0          4m13s
kube-system          kube-controller-manager-kind-control-plane   1/1     Running     0          4m15s
kube-system          kube-proxy-4rc5h                             1/1     Running     0          3m59s
kube-system          kube-scheduler-kind-control-plane            1/1     Running     0          4m13s
local-path-storage   local-path-provisioner-75f5b54ffd-t22jr      1/1     Running     0          3m59s
projectcontour       contour-68bdb8f5f8-644k4                     1/1     Running     0          3m59s
projectcontour       contour-68bdb8f5f8-tm9p8                     1/1     Running     0          3m59s
projectcontour       contour-certgen-v1.24.3-z6rcg                0/1     Completed   0          3m59s
projectcontour       envoy-84fxk                                  1/2     Running     0          3m56
```

**Comment by BenTheElder** [comment]
For contour maybe we should point to their own guide:

https://projectcontour.io/docs/1.24/guides/kind/

**Comment by brandonh-msft** [comment]
Thanks for the quick replies! I started over and didn't customize the ports, so I'm doing 80:8080. Now I get:

```
> docker exec -it kind-control-plane curl -v localhost/foo/echo?msg=foo
*   Trying 127.0.0.1:80...
* Connected to localhost (127.0.0.1) port 80 (#0)
> GET /foo/echo?msg=foo HTTP/1.1
> Host: localhost
> User-Agent: curl/7.81.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 404 Not Found
< vary: Accept-Encoding
< date: Fri, 21 Apr 2023 19:03:32 GMT
< server: envoy
< content-length: 0
<
* Connection #0 to host localhost left intact
```

**Comment by BenTheElder** [comment]
cc @sunjayBhatia

**Comment by brandonh-msft** [comment]
sorry - new to this. I see I should probably have hostport: 80, containerport: **80**... am re-working now.

**Comment by BenTheElder** [comment]
> Thanks for the quick replies! I started over and didn't customize the ports, so I'm doing 80:8080.

Ack thanks! Customizing the hostPort is fine and shouldn't be an issue, but it looks to me like contour is not working with kind or 1.27 currently.

We use the same port all the way through to minimize confusion, but you can use an arbitrary set of ports on the host, you just have to be able to match up things like if you change the hostPort in the kind config you need to use that port in the curl command or if you change the containerPort in the kind config then the ingress controller needs to be wired to that port as well.

In the meantime until we know what is wrong with contour I'd suggest using ingress-nginx + kind, which should work fine with the latest releases.

**Comment by BenTheElder** [comment]
Contour appears to be multiple versions behind kubernetes latest currently and we probably should not have it in the KIND docs as a result https://projectcontour.io/resources/compatibility-matrix/

KIND tracks Kubernetes closely as it's used to develop Kubernetes.

**Comment by brandonh-msft** [comment]
I can confirm switching just to NGINX and doing everything else the same, even using my original port #s, works as expected.

Thanks for your help!

**Comment by brandonh-msft** [comment]
Is it possible that having this in `usage.yaml` when _Contour_ is set up as ingress is causing a problem?
![image](https://user-images.githubusercontent.com/20270743/233722439-6a2bf05e-27df-48e5-9f68-7775a67be882.png)

**Comment by BenTheElder** [comment]
Contour should ignore the ingress-nginx annotations.

**Comment by BenTheElder** [comment]
kong was just fixed, but no updates for contour yet https://github.com/kubernetes-sigs/kind/pull/3183

**Comment by skriss** [comment]
Looks like, as @brandonh-msft suspected, in https://github.com/kubernetes-sigs/kind/pull/3025, the example Ingress was modified to use an nginx-specific annotation for doing path rewriting (`nginx.ingress.kubernetes.io/rewrite-target`), which the example relies on to function correctly (essentially stripping the `/foo/` and `/bar/` prefixes from the request paths so that just `/hostname` is sent to the backend services). Unfortunately, that nginx annotation is not supported by Contour.

Is anyone opposed to simplifying the example to just route to a single backend, so we can avoid implementation-specific issues like this?

**Comment by skriss** [comment]
I submitted https://github.com/kubernetes-sigs/kind/pull/3281, feedback welcome.
